internal/tutorial/measure/chapter07/main.go

```diff
@@ -0,0 +1,833 @@
+// -=-=- StartHere -=-=-
+//
+// # Chapter VII: Measuring a single URL
+//
+// This chapter is a digression. Rather than introducing you to
+// new functionality, we are going to put together everything we
+// have learned so far to write a simple URL scanner.
+//
+// Without further ado, let's describe our example `main.go` program
+// and let's use it to better understand this flow.
+//
+// (This file is auto-generated. Do not edit it directly! To apply
+// changes you need to modify `./internal/tutorial/measure/chapter07/main.go`.)
+//
+// ## main.go
+//
+// We have the declaration of the package and imports (as usual). However,
+// this program is more complex than previous programs, so we are going
+// to split it into separate functions. Let us start from the `main` function.
+//
+// ```Go
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/apex/log"
+	"github.com/miekg/dns"
+	"github.com/ooni/probe-cli/v3/internal/measure"
+	"github.com/ooni/probe-cli/v3/internal/runtimex"
+)
+
+// ```
+//
+// ### The main function
+//
+// ```Go
+func main() {
+	// ```
+	//
+	// The initial part of this program should look familiar. We are
+	// defining options using the `flag` package. Then we parse the
+	// command line flags using `Parse`. Then we create a context attached
+	// to a timeout for the whole scanning operation.
+	//
+	// ```Go
+	URL := flag.String("url", "https://dns.google/", "URL to measure")
+	timeout := flag.Duration("timeout", 30*time.Second, "timeout to use")
+	flag.Parse()
+	log.SetLevel(log.DebugLevel)
+	ctx, cancel := context.WithTimeout(context.Background(), *timeout)
+	defer cancel()
+	// ```
+	//
+	// Next we record the beginning of time, we create a new `Trace`,
+	// and we make an instance of the `urlMeasurer`. This type
+	// represents the possibility of measuring an URL (without any
+	// redirection for now).
+	//
+	// When we initialize the `urlMeasurer`, we pass to it an
+	// instance of `Measurer` created like we did in previous chapters.
+	//
+	// ```Go
+	begin := time.Now()
+	trace := measure.NewTrace(begin)
```

Comment:
TODO(bassosimone): using the same trace for all endpoints is wrong.

Reply:
Maybe it would make the code more robust to assign an `int64` ID to every connection, so that we can assert in the (future) tests that the network events of a given endpoint only contains a single connection.

Reply:
Done by refactoring and also created a TODO comment for the ID for every conn.